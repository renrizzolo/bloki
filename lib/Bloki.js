'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _BlokiProvider = require('./BlokiProvider');

var _reactEmotion = require('react-emotion');

var _reactEmotion2 = _interopRequireDefault(_reactEmotion);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Bloki = function (_Component) {
	_inherits(Bloki, _Component);

	function Bloki() {
		var _ref;

		var _temp, _this, _ret;

		_classCallCheck(this, Bloki);

		for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
			args[_key] = arguments[_key];
		}

		return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Bloki.__proto__ || Object.getPrototypeOf(Bloki)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
			computedStyle: {}
		}, _this.getWidth = function (width, columns, spacing) {
			var debug = _this.props.debug;

			var normalizedWidth = Math.max(0, Math.min(columns, width));

			debug && (console.log(normalizedWidth), console.log('calc(' + 100 / columns * normalizedWidth + '%)'));

			return Math.round(100 / columns * normalizedWidth) + '%';
		}, _this.columnWidth = function (width, theme, spacing) {
			var _this$props = _this.props,
			    col = _this$props.col,
			    row = _this$props.row,
			    auto = _this$props.auto,
			    nest = _this$props.nest;

			console.log('width:', width);
			if (col && width[theme.currentBreakpoint]) {
				return _this.getWidth(width[theme.currentBreakpoint], theme.columns, spacing);
			} else if (row) {
				return '100%';
			} else {
				return auto || !col && !row ? null : _this.getWidth(theme.columns, theme.columns, spacing);
			}
		}, _this.supports = function (property, value) {
			if ('CSS' in window && 'supports' in window.CSS) {
				return window.CSS.supports(property, value);
			}
		}, _this.getStyles = function (theme, colWidths) {
			var spacing = _this.props.spacing || theme.spacing || 0;
			var _this$props2 = _this.props,
			    col = _this$props2.col,
			    row = _this$props2.row,
			    auto = _this$props2.auto,
			    nest = _this$props2.nest,
			    sticky = _this$props2.sticky,
			    mb = _this$props2.mb,
			    innerSpacing = _this$props2.innerSpacing;
			var computedStyle = _this.state.computedStyle;

			var padding = void 0; //= `${spacing} ${spacing / 2}`;
			var flexGrow = 0;
			var margin = 0;

			if (col && row) {
				console.error('You must use either one of col or row (or neither for your container)');
			}

			if (col) {
				// margin = spacing / 2;
				padding = theme.innerSpacing ? spacing : 0;
				if (typeof innerSpacing === 'boolean') {
					padding = innerSpacing ? spacing : 0;
				}
			} else if (row) {
				padding = 0;
				margin = '-' + spacing + 'px';
			}

			if (col && auto || row) {
				flexGrow = 1;
			}
			if (nest) {
				margin = '-' + spacing / 2 + 'px';
			}

			var stickyPos = _this.supports('position', 'sticky') ? 'sticky' : '-webkit-sticky';
			var stickyStyle = sticky ? { top: 0, zIndex: 99, position: stickyPos } : {};

			var marginBottom = mb && typeof mb === 'boolean' ? spacing : mb;
			var styles = _extends({
				flexBasis: nest ? null : _this.columnWidth(colWidths, theme, spacing),
				maxWidth: nest ? null : _this.columnWidth(colWidths, theme, spacing),
				padding: padding,
				marginLeft: margin,
				marginRight: margin,
				marginBottom: marginBottom,
				flexGrow: flexGrow
			}, computedStyle, stickyStyle);
			_this.props.debug && console.log(styles);
			return styles;
		}, _this.getBreakpointUpStyle = function (currentBreakpoint, up) {
			// theme.up is an array of breakpoints includeing and above the current breakpoint
			// ['xs', 'md', 'lg', 'xl']
			var breakpointUpStyles = {};
			for (var i = 0; i < up.length; i++) {
				var styleKey = [up[i]] + 'UpStyle';
				_this.props.debug && console.log(styleKey + ':', _this.props[styleKey]);
				if (_this.props[styleKey]) {
					breakpointUpStyles = _extends({}, breakpointUpStyles, _this.props[styleKey]);
				}
			}
			return breakpointUpStyles;
		}, _this.Block = (0, _reactEmotion2.default)(_this.props.component === Bloki ? 'div' : _this.props.component)(function (props) {
			return _extends({}, props.styles);
		}), _temp), _possibleConstructorReturn(_this, _ret);
	}

	_createClass(Bloki, [{
		key: 'componentDidMount',
		value: function componentDidMount() {
			var _props = this.props,
			    wrap = _props.wrap,
			    align = _props.align,
			    justify = _props.justify,
			    col = _props.col,
			    row = _props.row,
			    spacing = _props.spacing,
			    sticky = _props.sticky,
			    style = _props.style;


			this.setState({
				computedStyle: {
					minHeight: '1px',
					position: 'relative',
					boxSizing: 'border-box',
					display: 'flex',
					flexWrap: wrap ? 'wrap' : 'nowrap',
					alignItems: align,
					justifyContent: justify,
					flexDirection: col ? 'column' : 'row'
				}
			});
		}
	}, {
		key: 'render',
		value: function render() {
			var _this2 = this;

			var _props2 = this.props,
			    children = _props2.children,
			    style = _props2.style,
			    debug = _props2.debug,
			    component = _props2.component,
			    auto = _props2.auto,
			    xs = _props2.xs,
			    sm = _props2.sm,
			    md = _props2.md,
			    lg = _props2.lg,
			    xl = _props2.xl,
			    rest = _objectWithoutProperties(_props2, ['children', 'style', 'debug', 'component', 'auto', 'xs', 'sm', 'md', 'lg', 'xl']);

			return _react2.default.createElement(
				_BlokiProvider.BlokiContext.Consumer,
				null,
				function (theme) {
					// breakpointStyle > style > props > theme
					if (!theme) {
						console.error('Bloki needs to be wrapped in a <BlokiProvider/>, with an optional theme');
						return;
					}

					// make the breakpoints above 
					// the first supplied equal to that
					// e.g xs:6 sm:8 md:null lg: null xl: null
					// becomes xs:6 sm:8 md:8 lg: 8 xl: 8
					//
					// unsupplied sizes below the lowest supplied
					//  are converted to max cols 
					// it's crude I know...

					var widths = { xl: xl, lg: lg, md: md, sm: sm, xs: xs };
					var arr = [xl, lg, md, sm, xs];

					Object.keys(widths).forEach(function (key, i) {
						console.log(arr[i], arr, i);
						if (!auto && !widths[key]) widths[key] = arr[i + 1] || arr[i + 2] || arr[i + 3] || arr[i + 3] || theme.columns;
					});
					console.log(widths);
					var styles = _this2.getStyles(theme, widths);
					var breakPointStyle = theme.currentBreakpoint + 'Style';
					var breakPointUpStyle = _this2.getBreakpointUpStyle(theme.currentBreakpoint, theme.up);
					return (
						// can be a render function returning the theme
						typeof children === 'function' ? _react2.default.createElement(
							_this2.Block,
							_extends({}, rest, {
								component: component,
								styles: _extends({}, styles, style, _this2.props[breakPointStyle], breakPointUpStyle)
							}),
							children(theme)
						) : _react2.default.createElement(
							_this2.Block,
							_extends({}, rest, {
								component: component,
								styles: _extends({}, styles, style, _this2.props[breakPointStyle], breakPointUpStyle)
							}),
							children
						)
					);
				}
			);
		}
	}]);

	return Bloki;
}(_react.Component);

exports.default = Bloki;


Bloki.propTypes = {
	component: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.function]),
	wrap: _propTypes2.default.bool,
	align: _propTypes2.default.string,
	justify: _propTypes2.default.string,
	innerSpacing: _propTypes2.default.bool,
	sticky: _propTypes2.default.bool,
	col: _propTypes2.default.bool,
	row: _propTypes2.default.bool,
	auto: _propTypes2.default.bool,
	nest: _propTypes2.default.bool,
	xs: _propTypes2.default.number,
	sm: _propTypes2.default.number,
	md: _propTypes2.default.number,
	lg: _propTypes2.default.number,
	xl: _propTypes2.default.number,
	mb: _propTypes2.default.oneOfType([_propTypes2.default.bool, _propTypes2.default.number]),
	xsUpStyle: _propTypes2.default.object,
	smUpStyle: _propTypes2.default.object,
	mdUpStyle: _propTypes2.default.object,
	lgUpStyle: _propTypes2.default.object,
	xlUpStyle: _propTypes2.default.object,
	debug: _propTypes2.default.bool
};
Bloki.defaultProps = {
	component: 'div',
	wrap: true,
	align: 'stretch',
	justify: 'flex-start',
	innerSpacing: null,
	sticky: false,
	col: false,
	row: false,
	auto: false,
	nest: false,
	xs: null,
	sm: null,
	md: null,
	lg: null,
	xl: null,
	mb: false,
	xsUpStyle: null,
	smUpStyle: null,
	mdUpStyle: null,
	lgUpStyle: null,
	xlUpStyle: null,
	debug: false
};