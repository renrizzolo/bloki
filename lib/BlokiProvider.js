'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.BlokiContext = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _util = require('./util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/*
	Bloki provider.
	Sets the breakpoint from the theme or default, and provides theme settings as context.
*/
var defaults = {
	spacing: 16,
	columns: 12,
	innerSpacing: true,
	breakpoints: {
		xs: 476,
		sm: 768,
		md: 992,
		lg: 1300
	}
};
var BlokiContext = exports.BlokiContext = _react2.default.createContext();

var BlokiProvider = function (_Component) {
	_inherits(BlokiProvider, _Component);

	function BlokiProvider() {
		var _ref;

		var _temp, _this, _ret;

		_classCallCheck(this, BlokiProvider);

		for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
			args[_key] = arguments[_key];
		}

		return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = BlokiProvider.__proto__ || Object.getPrototypeOf(BlokiProvider)).call.apply(_ref, [this].concat(args))), _this), _initialiseProps.call(_this), _temp), _possibleConstructorReturn(_this, _ret);
	}

	_createClass(BlokiProvider, [{
		key: 'componentWillMount',
		value: function componentWillMount() {
			this.setBreakpoint();
			this.setTheme();
		}
	}, {
		key: 'render',
		value: function render() {
			var children = this.props.children;
			var _state = this.state,
			    breakpoint = _state.breakpoint,
			    up = _state.up,
			    computedTheme = _state.computedTheme;

			return _react2.default.createElement(
				BlokiContext.Provider,
				{ value: _extends({}, computedTheme, { currentBreakpoint: breakpoint, up: up }) },
				children
			);
		}
	}]);

	return BlokiProvider;
}(_react.Component);

var _initialiseProps = function _initialiseProps() {
	var _this2 = this,
	    _arguments = arguments;

	this.state = {
		breakpoint: 'xl'
	};
	this.timeout = 200;

	this.componentDidMount = function () {
		_this2.listener = _this2.throttle(_this2.setBreakpoint, _this2.timeout);
		window.addEventListener('resize', _this2.listener);
	};

	this.componentWillUnmount = function () {
		window.removeEventListener('resize', _this2.listener);
	};

	this.throttle = function (func, timeout) {
		var inThrottle = void 0;
		return function () {
			var args = _arguments;
			var context = _this2;
			if (!inThrottle) {
				func.apply(context, args);
				inThrottle = true;
				setTimeout(function () {
					return inThrottle = false;
				}, timeout);
			}
		};
	};

	this.setBreakpoint = function () {
		var breakpoint = (0, _util.getBreakpoint)(_this2.props.theme.breakpoints);
		var breakpointsUpArray = ['xs', 'sm', 'md', 'lg', 'xl'];
		// return the breakpoints including and above the current
		var upArray = breakpointsUpArray.filter(function (item, index) {
			return index <= breakpointsUpArray.indexOf(breakpoint);
		});
		_this2.setState({
			breakpoint: breakpoint,
			up: upArray
		});
	};

	this.setTheme = function () {
		var theme = _this2.props.theme;

		var computedTheme = Object.assign({}, defaults || {}, theme);
		_this2.setState({ computedTheme: computedTheme });
	};
};

exports.default = BlokiProvider;


BlokiProvider.propTypes = {
	theme: _propTypes2.default.object,
	children: _propTypes2.default.node.isRequired
};
BlokiProvider.defaultProps = {
	theme: {}
};