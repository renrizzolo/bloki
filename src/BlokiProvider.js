import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { getBreakpoint } from './util';

/*
	Bloki provider.
	Sets the breakpoint from the theme or default, and provides theme settings as context.
*/
const defaults = {
		spacing: 16,
		columns: 12,
		innerSpacing: true,
		breakpoints: {
			xs: 476,
			sm: 768,
			md: 992,
			lg: 1300,
		}
	}
export const BlokiContext = React.createContext();

class BlokiProvider extends Component {
	state = {
		breakpoint: 'xl'
	}
	timeout = 125;
	componentWillMount() {
		this.setBreakpoint();
		this.setTheme();
	}
  componentDidMount = () => {
    this.listener = this.throttle(
    	this.setBreakpoint, this.timeout
    	);
    window.addEventListener('resize', this.listener);
  }

 	componentWillUnmount = () => {
    window.removeEventListener('resize', this.listener);
  }

	throttle = (func, timeout) => {
	  let inThrottle
	  return () => {
	    const args = arguments
	    const context = this
	    if (!inThrottle) {
	      func.apply(context, args)
	      inThrottle = true
	      setTimeout(() => inThrottle = false, timeout)
	    }
	  }
	}

	setBreakpoint = () => {
		const breakpoint = getBreakpoint(this.props.theme.breakpoints);
		const breakpointsUpArray = ['xs','sm','md','lg','xl'];
		// return the breakpoints including and above the current
		const upArray = breakpointsUpArray.filter((item, index) => index <= breakpointsUpArray.indexOf(breakpoint))
		this.setState({ 
			breakpoint: breakpoint,
			up: upArray
		});
	}
	setTheme = () => {
		const { theme } = this.props;
		const computedTheme = Object.assign({}, defaults || {},theme);
		this.setState({computedTheme})
	}
	render() {
		const { children } = this.props;
		const { breakpoint, up, computedTheme } = this.state;
		return (
				<BlokiContext.Provider value={{...computedTheme, currentBreakpoint: breakpoint, up: up}}>
					{children}
				</BlokiContext.Provider>
		);
	}
}

export default BlokiProvider;


BlokiProvider.propTypes = {
	theme: PropTypes.object,
	children: PropTypes.node.isRequired,
}
BlokiProvider.defaultProps =   {
	theme: {
	}
}

