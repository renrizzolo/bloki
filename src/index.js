import Bloki from './Bloki';
export BlokiProvider from './BlokiProvider';
// export BlokiProvider;
export default Bloki;