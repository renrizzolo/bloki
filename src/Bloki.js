import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { BlokiContext } from './BlokiProvider';

export default class Bloki extends Component {
	state = {
		computedStyle: {}
	}

	componentDidMount() {
		const {
			wrap,
			align,
			justify,
			col,
			row,
			spacing,
			sticky,
			style,
		} = this.props;
		
		this.setState({
			computedStyle: {
				minHeight: '1px',
    		position: 'relative',
				boxSizing: 'border-box',
				display: 'flex',
				flexWrap: wrap ? 'wrap' : 'nowrap',
				alignItems: align,
				justifyContent: justify,
				flexDirection: col ? 'column' : 'row',
			}
		})
	}

	getWidth = (width, columns, spacing) => {
	  const { debug } = this.props;
	  const normalizedWidth = Math.max(0, Math.min(columns, width));

	  debug && (
		 	console.log(normalizedWidth),
		 	console.log(`calc(${(100 / columns) * normalizedWidth}% - ${spacing}px)`)
		 );

	  return `calc(${(100 / columns) * normalizedWidth}% - ${spacing}px)`;
	};

	columnWidth = (width, theme, spacing) => {
		const { 			
			col,
			row,
			auto,
			nest
		} = this.props;

		if ( col && width[theme.currentBreakpoint] ) {
				return this.getWidth(width[theme.currentBreakpoint], theme.columns, spacing)
			} else if ( row ) {
				return '100%'
			} else {
				return auto || (!col && !row) ? null : this.getWidth(theme.columns, theme.columns, spacing)
			}
	}
	supports = (property, value) => {
  	if ('CSS' in window && 'supports' in window.CSS) {
    	return window.CSS.supports(property, value);
  	}
	}
	getStyles = (theme) => {
		const spacing = this.props.spacing || theme.spacing || 0;
		const { 			
			col,
			row,
			auto,
			nest,
			sticky,
			xs, sm,	md,	lg,	xl,
			mb,
			innerSpacing,
		} = this.props;
		const { computedStyle } = this.state;
		let padding;  //= `${spacing} ${spacing / 2}`;
		let flexGrow = 0;
		let margin = 0;

		if ( col && row ) {
			console.error('You must use either one of col or row (or neither for your container)')
		}

		if ( col ) {
			margin = spacing / 2;
			padding = theme.innerSpacing ? spacing : 0;
			if (typeof innerSpacing === 'boolean') {
				padding = innerSpacing ? spacing : 0;
			}
		}
		else if ( row ) {
			padding = 0;
			margin = `${spacing / 2}`;
		}

		if ( (col && auto) || row ) {
			flexGrow = 1;
		}
		if ( nest ) {
			margin = `-${spacing / 2}`;
		}

		const stickyPos = this.supports('position', 'sticky') ? 'sticky' : '-webkit-sticky';
		const stickyStyle = sticky ? { top: 0, zIndex: 99, position: stickyPos } : {};

		const marginBottom = mb && typeof mb === 'boolean' ? spacing : mb;
		const styles = {
    	flexBasis: nest ? null : this.columnWidth({xs, sm,	md,	lg,	xl}, theme, spacing),
      maxWidth: nest ? null : this.columnWidth({xs, sm,	md,	lg,	xl}, theme, spacing),
    	padding: padding,
    	marginLeft: margin,
			marginRight: margin,
			marginBottom: marginBottom,
			flexGrow: flexGrow,
    	...computedStyle,
    	...stickyStyle,
    }
    this.props.debug && console.log(styles);
		return styles;
	}
	getBreakpointUpStyle = (currentBreakpoint, up) => {
		// theme.up is an array of breakpoints includeing and above the current breakpoint
		// ['xs', 'md', 'lg', 'xl']
  	let breakpointUpStyles = {};
  	for (let i = 0; i < up.length; i++) {
  		const styleKey = `${[up[i]]}UpStyle`
  		this.props.debug && console.log(`${styleKey}:`, this.props[styleKey]);
  		if (this.props[styleKey]) {
  			breakpointUpStyles = {
  				...breakpointUpStyles, 
  				...this.props[styleKey]
  			}
  		}
  	}
    return breakpointUpStyles;
	}

	render() {
		const {
			children,
			style,
			debug,
			className,
			...rest
		} = this.props;
		
		return (
			<BlokiContext.Consumer>
      {(theme) => {
      	// breakpointStyle > style > props > theme
      	if (!theme) {
      		console.error('Bloki needs to be wrapped in a <BlokiProvider/>, with an optional theme')
      		return;
      	}
      	const styles = this.getStyles(theme);
      	const breakPointStyle = `${theme.currentBreakpoint}Style`;
      	const breakPointUpStyle = this.getBreakpointUpStyle(theme.currentBreakpoint, theme.up);
	      return (
	      // can be a render function returning the theme
	      typeof children === 'function' ?
	      <div 
		      className={className}
			    style={{...styles, ...style, ...this.props[breakPointStyle],  ...breakPointUpStyle}}
	      >
	    		{children(theme)}
	      </div>
	      :
	      <div
	      	className={className}
		      style={{...styles, ...style, ...this.props[breakPointStyle], ...breakPointUpStyle}}
	      >
					{children}
				</div>
			
				)
			}
		}
			</BlokiContext.Consumer>
		);
	}
}

Bloki.propTypes = {
	wrap: PropTypes.bool,
	align: PropTypes.string,
	justify: PropTypes.string,
	innerSpacing: PropTypes.bool,
	sticky: PropTypes.bool,
	col: PropTypes.bool,
	row: PropTypes.bool,
	auto: PropTypes.bool,
	nest: PropTypes.bool,
	xs: PropTypes.number,
	sm: PropTypes.number,
	md: PropTypes.number,
	lg: PropTypes.number,
	xl: PropTypes.number,
	mb: PropTypes.oneOfType([PropTypes.bool, PropTypes.number]),
	xsUpStyle: PropTypes.object,
	smUpStyle: PropTypes.object,
	mdUpStyle: PropTypes.object,
	lgUpStyle: PropTypes.object,
	xlUpStyle: PropTypes.object,
	debug: PropTypes.bool,
}
Bloki.defaultProps = {
	wrap: true,
	align: 'stretch',
	justify: 'flex-start',
	innerSpacing: null,
	sticky: false,
	col: false,
	row: false,
	auto: false,
	nest: false,
	xs: null,
	sm: null,
	md: null,
	lg: null,
	xl: null,
	mb: false,
	xsUpStyle: null,
	smUpStyle: null,
	mdUpStyle: null,
	lgUpStyle: null,
	xlUpStyle: null,
	debug: false,
}

