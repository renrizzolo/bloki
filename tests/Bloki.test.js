import React from 'react'
import { shallow, mount, render } from 'enzyme'
import Bloki, { BlokiProvider } from '../src';
import * as emotion from 'emotion'
import renderer from 'react-test-renderer'
import { createSerializer, createMatchers } from 'jest-emotion'
import { BlokiContext } from '../src/BlokiProvider';
expect.extend(createMatchers(emotion))

expect.addSnapshotSerializer(createSerializer(emotion))
// just a few basic tests

const defaultProps = { 
  wrap: true,
  align: 'stretch',
  justify: 'flex-start',
  innerSpacing: null,
  sticky: false,
  col: false,
  row: false,
  auto: false,
  nest: false,
  component: 'div',
  xs: null,
  sm: null,
  md: null,
  lg: null,
  xl: null,
  mb: false,
  xsUpStyle: null,
  smUpStyle: null,
  mdUpStyle: null,
  lgUpStyle: null,
  xlUpStyle: null,
  debug: false
}
const defaultTheme = {
  children: <Bloki/>,
  value:{
    spacing: 16,
    columns: 12,
    innerSpacing: true,
    breakpoints: {
      xs: 476,
      sm: 768,
      md: 992,
      lg: 1300,
    },
    currentBreakpoint: 'xs',
    up: [
      'xs',
    ]
  }
}

function createTestProps (props) {
  return {
    ...defaultProps,
    ...props,
  }
}

// force the window width to set the desired currentBreakpoint 
global.innerWidth = 475;


describe('rendering', () => {
  const props = createTestProps();
  const wrapper = shallow((
    <BlokiProvider>
      <Bloki {...props}>
      </Bloki>
    </BlokiProvider>
    ));
  const wrapperWithChild = shallow((
    <BlokiProvider>
      <Bloki>
        <div className="child"/>
      </Bloki>
    </BlokiProvider>
));

  describe('Bloki wrapper', () => {
    it('should render a Bloki', () => {
      expect(wrapper.find(Bloki)).toHaveLength(1);
    })
    it('should render children', () => {
      expect(wrapperWithChild.contains(<div className="child"/>)).toBe(true);
    });
    it('should have the default theme', () => {
      expect(wrapper.props()).toEqual(defaultTheme);
    }) 
    it('should have the default props', () => {
      expect(wrapper.find(Bloki).props()).toEqual(defaultProps);
    })
  })
})

describe('styles', () => {
  const spacing = 24;
  const props = createTestProps({
    justify: 'center',
    align: 'center',
    wrap: false,
    xs: 6,
    spacing,
    //auto: true,
    col: true
  });
  // force sm breakpoint
  global.innerWidth = 477;

  const tree = renderer.create(
    <BlokiProvider>
      <Bloki {...props}>
      </Bloki>
    </BlokiProvider>
  ).toJSON()
  describe('Bloki styles and properties', () => {
    it('Should match the snaphot', () => {
      expect(tree).toMatchSnapshot()
    })
    // it('Should apply auto styles', () => {
    //   expect(tree).not.toHaveStyleRule('max-width', '50%')
    // })
    it('Should apply wrap from props', () => {
      expect(tree).toHaveStyleRule('flex-wrap', 'nowrap')
    })
    it('Should have spacing applied to padding from props', () => {
      expect(tree).toHaveStyleRule('padding', `${spacing}px`)
    })
    // sm = 6
    it('Should have width applied based on column prop', () => {
      expect(tree).toHaveStyleRule('max-width', '50%')
    })
    it('Should have align-items applied based on align prop', () => {
      expect(tree).toHaveStyleRule('align-items', 'center')
    })   
    it('Should have justify-content applied based on justify prop', () => {
      expect(tree).toHaveStyleRule('justify-content', 'center')
    })
  })
})