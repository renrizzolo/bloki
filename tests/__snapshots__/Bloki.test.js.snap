// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`columns 1`] = `
.emotion-0 {
  padding: 16px;
  margin-left: 0;
  margin-right: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  min-height: 1px;
  position: relative;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
  className="emotion-0"
  wrap={true}
/>
`;

exports[`columns Bloki react emotion styles Should match the snaphot 1`] = `
.emotion-0 {
  padding: 16px;
  margin-left: 0;
  margin-right: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  min-height: 1px;
  position: relative;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
  className="emotion-0"
  wrap={true}
/>
`;

exports[`styles Bloki styles and properties Should match the snaphot 1`] = `
.emotion-0 {
  -webkit-flex-basis: 50%;
  -ms-flex-preferred-size: 50%;
  flex-basis: 50%;
  max-width: 50%;
  padding: 24px;
  margin-left: 0;
  margin-right: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  min-height: 1px;
  position: relative;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
  className="emotion-0"
  spacing={24}
  wrap={false}
/>
`;
